cat/ptx-proxy.cat

```diff
@@ -0,0 +1,81 @@
+PTX
+(* NVIDIA. Parallel Thread Execution ISA Version 7.5 *)
+
+(*******************)
+(*   Auxiliaries   *)
+(*******************)
+(* Events *)
+let strong-write = W & (RLX | REL)
+let strong-read = R & (RLX | ACQ)
+let strong-m = strong-write | strong-read
+let strong-operation = strong-write | strong-read | F
+
+(* Proxy *)
+let generic-proxy = GEN * GEN
+let surface-proxy = SUR * SUR
+let texture-proxy = TEX * TEX
+let constant-proxy = CON * CON
+let same-proxy = generic-proxy | surface-proxy | texture-proxy | constant-proxy
+
+(* Relations *)
+let sr-fence = ([strong-operation]; (sr & same-proxy); [F]) | ([F]; (sr & same-proxy); [strong-operation])
+let po-fence = ([strong-operation]; (po & same-proxy); [F]) | ([F]; (po & same-proxy); [strong-operation])
+let sr-m = ([strong-m]; (sr & loc & same-proxy); [strong-m])
+let po-m = ([M]; (po & loc & same-proxy); [M])
+let morally-strong = sr-fence | po-fence | sr-m | po-m
+(*******************)
+let fence-sc = [F & SC] ; sr; [F & SC]
+let rec observation = (morally-strong & rf) | (observation; rmw; observation)
+let release-pattern = ([W & REL]; po-loc?; [strong-write]) | ([F & ACQ_REL]; po; [strong-write])
+let acquire-pattern = ([strong-read]; po-loc?; [R & ACQ]) | ([strong-read]; po; [F & ACQ_REL])
+let sw = morally-strong & (release-pattern; observation+; acquire-pattern) | fence-sc
+let cause-base = (po*; (sw; po*)+) | po
```

Comment:
From this
```
fun sync : Op->Op {
  release_sequence.^observation.acquire_sequence
}

fun cause_base : Op->Op {
  ^po + *po.^((sc + sync).*po)
}
```
I would re-write this as follows
```
let sync = morally-strong & (release-pattern; observation+; acquire-pattern) 
let cause-base = (po*; ((sync | fence-sc); po*)+) | po
```
However their definition of `sync` does not have the `& morally strong`. Why did you add this?

Reply:
This should be another point that the alloy doen't agree with paper & [document](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#memory-synchronization):
"A release pattern X synchronizes with an acquire pattern Y, if a write operation in X precedes a read operation in Y in observation order, and the first operation in X and the last operation in Y are **morally strong**."
I think I'll add comments regarding this difference.

Reply:
The position of `fence-sc `is also different in paper and alloy.
`sw := (morally_strong ∩ (patternrel ; obs; patternacq )) ∪ syncbarrier ∪ sc`
 Both are the same I think, but the [document](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#memory-synchronization) seems to put it under the synchronize-with relation.

Reply:
I think we should try to stay as close to the Alloy model as possible (since we are suing this as our oracle to generate the expected results)

Reply:
Accepted